// @jbrowse/plugin-data-management/esm/AddTrackWidget/model.js
import { types } from 'mobx-state-tree';
import { ElementId } from '@jbrowse/core/util/types/mst';
import { guessAdapter, guessTrackType, getFileName, UNSUPPORTED, } from '@jbrowse/core/util/tracks';
function isAbsoluteUrl(url = '') {
    try {
        new URL(url);
        return true;
    }
    catch (error) {
        return url.startsWith('/');
    }
}
/**
 * #stateModel AddTrackModel
 */
export default function f(pluginManager) {
    return types
        .model('AddTrackModel', {
        /**
         * #property
         */
        id: ElementId,
        /**
         * #property
         */
        type: types.literal('AddTrackWidget'),
        /**
         * #property
         */
        view: types.safeReference(pluginManager.pluggableMstType('view', 'stateModel')),
    })
        .volatile(() => ({
        trackSource: 'fromFile',
        trackData: undefined,
        indexTrackData: undefined,
        // alts
        altAssemblyName: '',
        altTrackName: '',
        altTrackType: '',
        adapterHint: '',
        textIndexTrack: true,
        textIndexingConf: undefined,
    }))
        .actions(self => ({
        /**
         * #action
         */
        setAdapterHint(obj) {
            self.adapterHint = obj;
        },
        /**
         * #action
         */
        setTrackSource(str) {
            self.trackSource = str;
        },
        /**
         * #action
         */
        setTextIndexingConf(conf) {
            self.textIndexingConf = conf;
        },
        /**
         * #action
         */
        setTextIndexTrack(flag) {
            self.textIndexTrack = flag;
        },
        /**
         * #action
         */
        setTrackData(obj) {
            self.trackData = obj;
        },
        /**
         * #action
         */
        setIndexTrackData(obj) {
            self.indexTrackData = obj;
        },
        /**
         * #action
         */
        setAssembly(str) {
            self.altAssemblyName = str;
        },
        /**
         * #action
         */
        setTrackName(str) {
            self.altTrackName = str;
        },
        /**
         * #action
         */
        setTrackType(str) {
            self.altTrackType = str;
        },
        /**
         * #action
         */
        clearData() {
            self.trackSource = '';
            self.altTrackName = '';
            self.altTrackType = '';
            self.altAssemblyName = '';
            self.adapterHint = '';
            self.indexTrackData = undefined;
            self.trackData = undefined;
            self.textIndexingConf = undefined;
            self.textIndexTrack = true;
        },
    }))
        .views(self => ({
        /**
         * #getter
         */
        get trackAdapter() {
            const { trackData, indexTrackData, adapterHint } = self;
            return trackData
                ? guessAdapter(trackData, indexTrackData, adapterHint, self)
                : undefined;
        },
        /**
         * #getter
         */
        get trackName() {
            return (self.altTrackName ||
                (self.trackData ? getFileName(self.trackData) : ''));
        },
        /**
         * #getter
         */
        get isFtp() {
            var _a, _b;
            const { trackData: track, indexTrackData: index } = self;
            return !!(
            // @ts-expect-error
            (((_a = index === null || index === void 0 ? void 0 : index.uri) === null || _a === void 0 ? void 0 : _a.startsWith('ftp://')) || ((_b = track === null || track === void 0 ? void 0 : track.uri) === null || _b === void 0 ? void 0 : _b.startsWith('ftp://'))));
        },
        /**
         * #getter
         */
        get isRelativeTrackUrl() {
            var _a;
            // @ts-expect-error
            const uri = (_a = self.trackData) === null || _a === void 0 ? void 0 : _a.uri;
            return uri ? !isAbsoluteUrl(uri) : false;
        },
        /**
         * #getter
         */
        get isRelativeIndexUrl() {
            var _a;
            // @ts-expect-error
            const uri = (_a = self.indexTrackData) === null || _a === void 0 ? void 0 : _a.uri;
            return uri ? !isAbsoluteUrl(uri) : false;
        },
        /**
         * #getter
         */
        get isRelativeUrl() {
            return this.isRelativeIndexUrl || this.isRelativeTrackUrl;
        },
        /**
         * #getter
         */
        get trackHttp() {
            var _a, _b;
            // @ts-expect-error
            return (_b = (_a = self.trackData) === null || _a === void 0 ? void 0 : _a.uri) === null || _b === void 0 ? void 0 : _b.startsWith('http://');
        },
        /**
         * #getter
         */
        get indexHttp() {
            var _a, _b;
            // @ts-expect-error
            return (_b = (_a = self.indexTrackData) === null || _a === void 0 ? void 0 : _a.uri) === null || _b === void 0 ? void 0 : _b.startsWith('http://');
        },
        /**
         * #getter
         */
        get wrongProtocol() {
            return (window.location.protocol === 'https:' &&
                (this.trackHttp || this.indexHttp));
        },
        /**
         * #getter
         */
        get unsupported() {
            var _a;
            return ((_a = this.trackAdapter) === null || _a === void 0 ? void 0 : _a.type) === UNSUPPORTED;
        },
        /**
         * #getter
         */
        get assembly() {
            var _a;
            return self.altAssemblyName || ((_a = self.view.assemblyNames) === null || _a === void 0 ? void 0 : _a[0]);
        },
        /**
         * #getter
         */
        get trackType() {
            return (self.altTrackType ||
                (this.trackAdapter
                    ? guessTrackType(this.trackAdapter.type, self)
                    : ''));
        },
    }))
        .views(self => ({
        /**
         * #getter
         */
        get warningMessage() {
            if (self.isFtp) {
                return `Warning: JBrowse cannot access files using the ftp protocol`;
            }
            else if (self.isRelativeUrl) {
                return `Warning: one or more of your files do not provide the protocol e.g.
          https://, please provide an absolute URL unless you are sure a
          relative URL is intended.`;
            }
            else if (self.wrongProtocol) {
                return `Warning: You entered a http:// resources but we cannot access HTTP
          resources from JBrowse when it is running on https. Please use an
          https URL for your track, or access the JBrowse app from the http
          protocol`;
            }
            return '';
        },
    }));
}
