// @jbrowse/plugin-wiggle/esm/MultiLinearWiggleDisplay/components/DraggableDialog.js
import React from 'react';
import { Dialog, DialogTitle, IconButton, Divider, Paper, ScopedCssBaseline, } from '@mui/material';
import { observer } from 'mobx-react';
import { makeStyles } from 'tss-react/mui';
import Draggable from 'react-draggable';
// icons
import CloseIcon from '@mui/icons-material/Close';
const useStyles = makeStyles()(theme => ({
    closeButton: {
        position: 'absolute',
        right: theme.spacing(1),
        top: theme.spacing(1),
        color: theme.palette.grey[500],
    },
}));
// draggable dialog demo https://mui.com/material-ui/react-dialog/#draggable-dialog
function PaperComponent(props) {
    return (React.createElement(Draggable, { handle: "#draggable-dialog-title", cancel: '[class*="MuiDialogContent-root"]' },
        React.createElement(Paper, { ...props })));
}
const DraggableDialog = observer(function DraggableDialog(props) {
    const { classes } = useStyles();
    const { title, children, onClose } = props;
    return (React.createElement(Dialog, { ...props, PaperComponent: PaperComponent, "aria-labelledby": "draggable-dialog-title" // this area is important for the draggable functionality
     },
        React.createElement(ScopedCssBaseline, null,
            React.createElement(DialogTitle, { style: { cursor: 'move' }, id: "draggable-dialog-title" },
                title,
                onClose ? (React.createElement(IconButton, { className: classes.closeButton, onClick: () => {
                        // @ts-expect-error
                        onClose();
                    } },
                    React.createElement(CloseIcon, null))) : null),
            React.createElement(Divider, null),
            children)));
});
export default DraggableDialog;
