// @gmod/tabix/esm/indexFile.js
export default class IndexFile {
    constructor({ filehandle, renameRefSeqs = (n) => n, }) {
        this.filehandle = filehandle;
        this.renameRefSeq = renameRefSeqs;
    }
    async getMetadata(opts = {}) {
        const { indices: _indices, ...rest } = await this.parse(opts);
        return rest;
    }
    _findFirstData(currentFdl, virtualOffset) {
        if (currentFdl) {
            return currentFdl.compareTo(virtualOffset) > 0
                ? virtualOffset
                : currentFdl;
        }
        else {
            return virtualOffset;
        }
    }
    async parse(opts = {}) {
        if (!this.parseP) {
            this.parseP = this._parse(opts).catch(e => {
                this.parseP = undefined;
                throw e;
            });
        }
        return this.parseP;
    }
    async hasRefSeq(seqId, opts = {}) {
        var _a;
        const idx = await this.parse(opts);
        return !!((_a = idx.indices[seqId]) === null || _a === void 0 ? void 0 : _a.binIndex);
    }
}
