// victory-bar/lib/bar-helper-methods.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getStyle = exports.getCornerRadius = exports.getBarWidth = void 0;

var _isPlainObject2 = _interopRequireDefault(require("lodash/isPlainObject"));

var _victoryCore = require("victory-core");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DEFAULT_BAR_WIDTH = 8;

var getBarWidth = function (barWidth, props) {
  var scale = props.scale,
      data = props.data,
      style = props.style;

  if (barWidth) {
    return _victoryCore.Helpers.evaluateProp(barWidth, props);
  } else if (style.width) {
    return style.width;
  }

  var range = scale.x.range();
  var extent = Math.abs(range[1] - range[0]);
  var bars = data.length + 2;
  var barRatio = props.barRatio || 0.5;
  var defaultWidth = barRatio * (data.length < 2 ? DEFAULT_BAR_WIDTH : extent / bars);
  return Math.max(1, defaultWidth);
};

exports.getBarWidth = getBarWidth;

var getCornerRadiusFromObject = function (cornerRadius, props) {
  var realCornerRadius = {
    topLeft: 0,
    topRight: 0,
    bottomLeft: 0,
    bottomRight: 0
  };

  var updateCornerRadius = function (corner, fallback) {
    if (!_victoryCore.Helpers.isNil(cornerRadius[corner])) {
      realCornerRadius[corner] = _victoryCore.Helpers.evaluateProp(cornerRadius[corner], props);
    } else if (!_victoryCore.Helpers.isNil(cornerRadius[fallback])) {
      realCornerRadius[corner] = _victoryCore.Helpers.evaluateProp(cornerRadius[fallback], props);
    }
  };

  updateCornerRadius("topLeft", "top");
  updateCornerRadius("topRight", "top");
  updateCornerRadius("bottomLeft", "bottom");
  updateCornerRadius("bottomRight", "bottom");
  return realCornerRadius;
};

function isCornerRadiusObject(cornerRadius) {
  return (0, _isPlainObject2.default)(cornerRadius);
}

var getCornerRadius = function (cornerRadius, props) {
  var realCornerRadius = {
    topLeft: 0,
    topRight: 0,
    bottomLeft: 0,
    bottomRight: 0
  };

  if (!cornerRadius) {
    return realCornerRadius;
  }

  if (isCornerRadiusObject(cornerRadius)) {
    return getCornerRadiusFromObject(cornerRadius, props);
  }

  realCornerRadius.topLeft = _victoryCore.Helpers.evaluateProp(cornerRadius, props);
  realCornerRadius.topRight = _victoryCore.Helpers.evaluateProp(cornerRadius, props);
  return realCornerRadius;
};

exports.getCornerRadius = getCornerRadius;

var getStyle = function () {
  var style = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var props = arguments.length > 1 ? arguments[1] : undefined;

  if (props.disableInlineStyles) {
    return {};
  }

  var stroke = style.fill || "black";
  var baseStyle = {
    fill: "black",
    stroke: stroke
  };
  return _victoryCore.Helpers.evaluateStyle(Object.assign(baseStyle, style), props);
};

exports.getStyle = getStyle;
