// @jbrowse/plugin-linear-genome-view/esm/BaseLinearDisplay/models/TrackHeightMixin.js
import { getConf } from '@jbrowse/core/configuration';
import { types } from 'mobx-state-tree';
const minDisplayHeight = 20;
/**
 * #stateModel TrackHeightMixin
 * #category display
 */
export default function TrackHeightMixin() {
    return types
        .model({
        /**
         * #property
         */
        heightPreConfig: types.maybe(types.refinement('displayHeight', types.number, n => n >= minDisplayHeight)),
    })
        .volatile(() => ({
        /**
         * #property
         */
        scrollTop: 0,
    }))
        .views(self => ({
        get height() {
            var _a;
            // @ts-expect-error
            return (_a = self.heightPreConfig) !== null && _a !== void 0 ? _a : getConf(self, 'height');
        },
    }))
        .actions(self => ({
        /**
         * #action
         */
        setScrollTop(scrollTop) {
            self.scrollTop = scrollTop;
        },
        /**
         * #action
         */
        setHeight(displayHeight) {
            self.heightPreConfig = Math.max(displayHeight, minDisplayHeight);
            return self.height;
        },
        /**
         * #action
         */
        resizeHeight(distance) {
            const oldHeight = self.height;
            const newHeight = this.setHeight(self.height + distance);
            return newHeight - oldHeight;
        },
    }));
}
