// @jbrowse/plugin-alignments/esm/PileupRenderer/components/PileupRendering.js
import React, { useRef, useState, useEffect } from 'react';
import { PrerenderedCanvas } from '@jbrowse/core/ui';
import { bpSpanPx } from '@jbrowse/core/util';
import { observer } from 'mobx-react';
// used so that user can click-away-from-feature below the laid out features
// (issue #1248)
const canvasPadding = 100;
const PileupRendering = observer(function (props) {
    var _a;
    const { onMouseMove, blockKey, displayModel, width, height, regions, bpPerPx, sortedBy, colorBy, filterBy, } = props;
    const { selectedFeatureId, featureIdUnderMouse, contextMenuFeature } = displayModel;
    const [region] = regions;
    const highlightOverlayCanvas = useRef(null);
    const [mouseIsDown, setMouseIsDown] = useState(false);
    const [movedDuringLastMouseDown, setMovedDuringLastMouseDown] = useState(false);
    useEffect(() => {
        var _a, _b;
        const canvas = highlightOverlayCanvas.current;
        if (!canvas) {
            return;
        }
        const ctx = canvas.getContext('2d');
        if (!ctx) {
            return;
        }
        ctx.clearRect(0, 0, canvas.width, canvas.height);
        const selectedRect = selectedFeatureId
            ? (_a = displayModel.getFeatureByID) === null || _a === void 0 ? void 0 : _a.call(displayModel, blockKey, selectedFeatureId)
            : undefined;
        if (selectedRect) {
            const [leftBp, topPx, rightBp, bottomPx] = selectedRect;
            const [leftPx, rightPx] = bpSpanPx(leftBp, rightBp, region, bpPerPx);
            const rectTop = Math.round(topPx);
            const rectHeight = Math.round(bottomPx - topPx);
            ctx.shadowColor = '#222266';
            ctx.shadowBlur = 10;
            ctx.lineJoin = 'bevel';
            ctx.lineWidth = 2;
            ctx.strokeStyle = '#00b8ff';
            ctx.strokeRect(leftPx - 2, rectTop - 2, rightPx - leftPx + 4, rectHeight + 4);
            ctx.clearRect(leftPx, rectTop, rightPx - leftPx, rectHeight);
        }
        const highlightedFeature = featureIdUnderMouse || (contextMenuFeature === null || contextMenuFeature === void 0 ? void 0 : contextMenuFeature.id());
        const highlightedRect = highlightedFeature
            ? (_b = displayModel.getFeatureByID) === null || _b === void 0 ? void 0 : _b.call(displayModel, blockKey, highlightedFeature)
            : undefined;
        if (highlightedRect) {
            const [leftBp, topPx, rightBp, bottomPx] = highlightedRect;
            const [leftPx, rightPx] = bpSpanPx(leftBp, rightBp, region, bpPerPx);
            const rectTop = Math.round(topPx);
            const rectHeight = Math.round(bottomPx - topPx);
            ctx.fillStyle = '#0003';
            ctx.fillRect(leftPx, rectTop, rightPx - leftPx, rectHeight);
        }
    }, [
        bpPerPx,
        region,
        blockKey,
        selectedFeatureId,
        displayModel,
        featureIdUnderMouse,
        contextMenuFeature,
    ]);
    function onMouseDown(event) {
        setMouseIsDown(true);
        setMovedDuringLastMouseDown(false);
        callMouseHandler('MouseDown', event);
    }
    function onMouseEnter(event) {
        callMouseHandler('MouseEnter', event);
    }
    function onMouseOut(event) {
        callMouseHandler('MouseOut', event);
        callMouseHandler('MouseLeave', event);
    }
    function onMouseOver(event) {
        callMouseHandler('MouseOver', event);
    }
    function onMouseUp(event) {
        setMouseIsDown(false);
        callMouseHandler('MouseUp', event);
    }
    function onClick(event) {
        if (!movedDuringLastMouseDown) {
            callMouseHandler('Click', event);
        }
    }
    function onMouseLeave(event) {
        callMouseHandler('MouseOut', event);
        callMouseHandler('MouseLeave', event);
    }
    function onContextMenu(event) {
        callMouseHandler('ContextMenu', event);
    }
    function mouseMove(event) {
        if (!highlightOverlayCanvas.current) {
            return;
        }
        if (mouseIsDown) {
            setMovedDuringLastMouseDown(true);
        }
        const rect = highlightOverlayCanvas.current.getBoundingClientRect();
        const offsetX = event.clientX - rect.left;
        const offsetY = event.clientY - rect.top;
        const px = region.reversed ? width - offsetX : offsetX;
        const clientBp = region.start + bpPerPx * px;
        onMouseMove === null || onMouseMove === void 0 ? void 0 : onMouseMove(event, displayModel.getFeatureOverlapping(blockKey, clientBp, offsetY));
    }
    function callMouseHandler(handlerName, event) {
        // @ts-expect-error
        const featureHandler = props[`onFeature${handlerName}`];
        // @ts-expect-error
        const canvasHandler = props[`on${handlerName}`];
        if (featureHandler && featureIdUnderMouse) {
            featureHandler(event, featureIdUnderMouse);
        }
        else if (canvasHandler) {
            canvasHandler(event, featureIdUnderMouse);
        }
    }
    const canvasWidth = Math.ceil(width);
    // need to call this in render so we get the right observer behavior
    return (React.createElement("div", { "data-testid": `pileup-${[
            sortedBy === null || sortedBy === void 0 ? void 0 : sortedBy.type,
            colorBy === null || colorBy === void 0 ? void 0 : colorBy.type,
            colorBy === null || colorBy === void 0 ? void 0 : colorBy.tag,
            (_a = filterBy === null || filterBy === void 0 ? void 0 : filterBy.tagFilter) === null || _a === void 0 ? void 0 : _a.tag,
        ]
            .filter(f => !!f)
            .join('-')}`, style: { position: 'relative', width: canvasWidth, height } },
        React.createElement(PrerenderedCanvas, { ...props, style: { position: 'absolute', left: 0, top: 0 } }),
        React.createElement("canvas", { "data-testid": "pileup_overlay_canvas", width: canvasWidth, height: height + canvasPadding, style: { position: 'absolute', left: 0, top: 0 }, className: "highlightOverlayCanvas", ref: highlightOverlayCanvas, onMouseDown: event => onMouseDown(event), onMouseEnter: event => onMouseEnter(event), onMouseOut: event => onMouseOut(event), onMouseOver: event => onMouseOver(event), onMouseUp: event => onMouseUp(event), onMouseLeave: event => onMouseLeave(event), onMouseMove: event => mouseMove(event), onClick: event => onClick(event), onContextMenu: event => onContextMenu(event), onFocus: () => { }, onBlur: () => { } })));
});
export default PileupRendering;
