// @jbrowse/core/ui/colors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ggplot2Colors3 = exports.ggplot2Colors4 = exports.ggplot2Colors5 = exports.ggplot2Colors6 = exports.category10 = exports.dark2 = exports.tableau10 = exports.set2 = exports.set1 = void 0;
const category10 = [
    '#1f77b4',
    '#ff7f0e',
    '#2ca02c',
    '#d62728',
    '#9467bd',
    '#8c564b',
    '#e377c2',
    '#7f7f7f',
    '#bcbd22',
    '#17becf',
];
exports.category10 = category10;
const dark2 = [
    '#1b9e77',
    '#d95f02',
    '#7570b3',
    '#e7298a',
    '#66a61e',
    '#e6ab02',
    '#a6761d',
    '#666666',
];
exports.dark2 = dark2;
const set1 = [
    '#e41a1c',
    '#377eb8',
    '#4daf4a',
    '#984ea3',
    '#ff7f00',
    '#ffff33',
    '#a65628',
    '#f781bf',
    '#999999',
];
exports.set1 = set1;
const set2 = [
    '#66c2a5',
    '#fc8d62',
    '#8da0cb',
    '#e78ac3',
    '#a6d854',
    '#ffd92f',
    '#e5c494',
    '#b3b3b3',
];
exports.set2 = set2;
const tableau10 = [
    '#4e79a7',
    '#f28e2c',
    '#e15759',
    '#76b7b2',
    '#59a14f',
    '#edc949',
    '#af7aa1',
    '#ff9da7',
    '#9c755f',
    '#bab0ab',
];
exports.tableau10 = tableau10;
const ggplot2Colors6 = [
    '#F8766D',
    '#B79F00',
    '#00BA38',
    '#00BFC4',
    '#619CFF',
    '#F564E3',
];
exports.ggplot2Colors6 = ggplot2Colors6;
const ggplot2Colors5 = ['#F8766D', '#A3A500', '#00BF7D', '#00B0F6', '#E76BF3'];
exports.ggplot2Colors5 = ggplot2Colors5;
const ggplot2Colors4 = ['#F8766D', '#7CAE00', '#00BFC4', '#C77CFF'];
exports.ggplot2Colors4 = ggplot2Colors4;
const ggplot2Colors3 = ['#F8766D', '#00BA38', '#619CFF'];
exports.ggplot2Colors3 = ggplot2Colors3;
