// victory-legend/es/victory-legend.js
function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (typeof call === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

import React from "react";
import { getBaseProps, getDimensions } from "./helper-methods";
import { addEvents, Helpers, VictoryLabel, VictoryContainer, VictoryTheme, Point, Border } from "victory-core";
var fallbackProps = {
  orientation: "vertical",
  titleOrientation: "top",
  width: 450,
  height: 300,
  x: 0,
  y: 0
};
var defaultLegendData = [{
  name: "Series 1"
}, {
  name: "Series 2"
}]; // eslint-disable-next-line @typescript-eslint/no-empty-interface

var VictoryLegendBase = /*#__PURE__*/function (_React$Component) {
  _inherits(VictoryLegendBase, _React$Component);

  var _super = _createSuper(VictoryLegendBase);

  function VictoryLegendBase() {
    _classCallCheck(this, VictoryLegendBase);

    return _super.apply(this, arguments);
  }

  _createClass(VictoryLegendBase, [{
    key: "renderChildren",
    value: function renderChildren(props) {
      var _this = this;

      var dataComponent = props.dataComponent,
          labelComponent = props.labelComponent,
          title = props.title;
      var children = [];

      if (props.borderComponent) {
        var borderProps = this.getComponentProps(props.borderComponent, "border", "all");
        var borderComponent = /*#__PURE__*/React.cloneElement(props.borderComponent, borderProps);
        children.push(borderComponent);
      }

      if (dataComponent) {
        var dataComponents = this.dataKeys.map(function (_dataKey, index) {
          if (_dataKey === "all") {
            return undefined;
          }

          var dataProps = _this.getComponentProps(dataComponent, "data", index);

          return /*#__PURE__*/React.cloneElement(dataComponent, dataProps);
        }).filter(function (comp) {
          return comp !== undefined;
        });
        children.push.apply(children, _toConsumableArray(dataComponents));
      }

      if (title && props.titleComponent) {
        var titleProps = this.getComponentProps(title, "title", "all");
        var titleComponent = /*#__PURE__*/React.cloneElement(props.titleComponent, titleProps);
        children.push(titleComponent);
      }

      if (labelComponent) {
        var labelComponents = this.dataKeys.map(function (_dataKey, index) {
          if (_dataKey === "all") {
            return undefined;
          }

          var labelProps = _this.getComponentProps(labelComponent, "labels", index);

          if (labelProps.text !== undefined && labelProps.text !== null) {
            return /*#__PURE__*/React.cloneElement(labelComponent, labelProps);
          }

          return undefined;
        }).filter(function (comp) {
          return comp !== undefined;
        });
        children.push.apply(children, _toConsumableArray(labelComponents));
      }

      return children;
    }
  }, {
    key: "render",
    value: function render() {
      // @ts-expect-error Property 'role' does not exist on type 'Function'.
      // Ref: https://github.com/microsoft/TypeScript/issues/32452
      var role = this.constructor.role;
      var props = Helpers.modifyProps(this.props, fallbackProps, role);
      var children = this.renderChildren(props);
      return props.standalone ? this.renderContainer(props.containerComponent, children) : /*#__PURE__*/React.cloneElement(props.groupComponent, {}, children);
    }
  }]);

  return VictoryLegendBase;
}(React.Component);

VictoryLegendBase.displayName = "VictoryLegend";
VictoryLegendBase.role = "legend";
VictoryLegendBase.defaultProps = {
  borderComponent: /*#__PURE__*/React.createElement(Border, null),
  data: defaultLegendData,
  containerComponent: /*#__PURE__*/React.createElement(VictoryContainer, null),
  dataComponent: /*#__PURE__*/React.createElement(Point, null),
  groupComponent: /*#__PURE__*/React.createElement("g", null),
  labelComponent: /*#__PURE__*/React.createElement(VictoryLabel, null),
  standalone: true,
  theme: VictoryTheme.grayscale,
  titleComponent: /*#__PURE__*/React.createElement(VictoryLabel, null)
};

VictoryLegendBase.getBaseProps = function (props) {
  return getBaseProps(props, fallbackProps);
};

VictoryLegendBase.getDimensions = function (props) {
  return getDimensions(props, fallbackProps);
};

VictoryLegendBase.expectedComponents = ["borderComponent", "containerComponent", "dataComponent", "groupComponent", "labelComponent", "titleComponent"];
export var VictoryLegend = addEvents(VictoryLegendBase);
