// @jbrowse/plugin-legacy-jbrowse/esm/NCListAdapter/NCListAdapter.js
import NCListStore from '@gmod/nclist';
import { BaseFeatureDataAdapter, } from '@jbrowse/core/data_adapters/BaseAdapter';
import { ObservableCreate } from '@jbrowse/core/util/rxjs';
import { checkAbortSignal } from '@jbrowse/core/util';
import { RemoteFile } from 'generic-filehandle';
import NCListFeature from './NCListFeature';
export default class NCListAdapter extends BaseFeatureDataAdapter {
    constructor(config, getSubAdapter, pluginManager) {
        super(config, getSubAdapter, pluginManager);
        const refNames = this.getConf('refNames');
        const rootUrlTemplate = this.getConf('rootUrlTemplate');
        this.configRefNames = refNames;
        this.nclist = new NCListStore({
            baseUrl: '',
            urlTemplate: rootUrlTemplate.uri,
            readFile: (url) => new RemoteFile(String(rootUrlTemplate.baseUri
                ? new URL(url, rootUrlTemplate.baseUri).toString()
                : url)).readFile(),
        });
    }
    /**
     * Fetch features for a certain region. Use getFeaturesInRegion() if you also
     * want to verify that the store has features for the given reference sequence
     * before fetching.
     * @param region -
     * @param opts - [signal] optional signalling object for aborting the fetch
     * @returns Observable of Feature objects in the region
     */
    getFeatures(region, opts = {}) {
        return ObservableCreate(async (observer) => {
            const { signal } = opts;
            for await (const feature of this.nclist.getFeatures(region, opts)) {
                checkAbortSignal(signal);
                observer.next(this.wrapFeature(feature));
            }
            observer.complete();
        });
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    wrapFeature(ncFeature) {
        return new NCListFeature(ncFeature, undefined, `${this.id}-${ncFeature.id()}`);
    }
    async hasDataForRefName(refName) {
        var _a;
        const root = await this.nclist.getDataRoot(refName);
        return !!((_a = root === null || root === void 0 ? void 0 : root.stats) === null || _a === void 0 ? void 0 : _a.featureCount);
    }
    /**
     * NCList is unable to get list of ref names so returns empty
     */
    async getRefNames() {
        return this.configRefNames || [];
    }
    /**
     * called to provide a hint that data tied to a certain region
     * will not be needed for the foreseeable future and can be purged
     * from caches, etc
     */
    freeResources() { }
}
