// victory-voronoi/lib/voronoi.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Voronoi = void 0;

var _react = _interopRequireDefault(require("react"));

var _victoryCore = require("victory-core");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var getVoronoiPath = function (props) {
  var _props$polygon;

  var polygon = props.polygon;
  return Array.isArray(polygon) && polygon.length ? "M ".concat((_props$polygon = props.polygon) === null || _props$polygon === void 0 ? void 0 : _props$polygon.join("L"), " Z") : "";
};

function evaluateProps(props) {
  /**
   * Potential evaluated props are:
   * `aria-label`
   * `id`
   * `size`
   * `style`
   * `tabIndex
   */
  var ariaLabel = _victoryCore.Helpers.evaluateProp(props.ariaLabel, props);

  var id = _victoryCore.Helpers.evaluateProp(props.id, props);

  var size = _victoryCore.Helpers.evaluateProp(props.size, props);

  var style = _victoryCore.Helpers.evaluateStyle(props.style, props);

  var tabIndex = _victoryCore.Helpers.evaluateProp(props.tabIndex, props);

  return Object.assign({}, props, {
    ariaLabel: ariaLabel,
    id: id,
    size: size,
    style: style,
    tabIndex: tabIndex
  });
}

var defaultProps = {
  pathComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.Path, null),
  circleComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.Circle, null),
  clipPathComponent: /*#__PURE__*/_react.default.createElement(_victoryCore.ClipPath, null),
  groupComponent: /*#__PURE__*/_react.default.createElement("g", null),
  role: "presentation",
  shapeRendering: "auto"
};

var Voronoi = function (initialProps) {
  var props = evaluateProps(_objectSpread(_objectSpread({}, defaultProps), initialProps));
  var ariaLabel = props.ariaLabel,
      role = props.role,
      shapeRendering = props.shapeRendering,
      className = props.className,
      events = props.events,
      transform = props.transform,
      style = props.style,
      size = props.size,
      tabIndex = props.tabIndex;
  var voronoiPath = getVoronoiPath(props);

  var sharedProps = _objectSpread({
    "aria-label": ariaLabel,
    className: className,
    role: role,
    shapeRendering: shapeRendering,
    style: style,
    tabIndex: tabIndex,
    transform: transform
  }, events);

  var userProps = _victoryCore.UserProps.getSafeUserProps(props);

  if (size) {
    var circle = /*#__PURE__*/_react.default.cloneElement(props.circleComponent, _objectSpread(_objectSpread({}, sharedProps), {}, {
      key: "".concat(props.id, "-circle-clip"),
      clipPath: "url(#".concat(props.clipId, ")"),
      cx: props.x,
      cy: props.y,
      r: size
    }));

    var voronoiClipPath = /*#__PURE__*/_react.default.cloneElement(props.clipPathComponent, {
      key: "".concat(props.id, "-voronoi-clip"),
      clipId: props.clipId
    }, /*#__PURE__*/_react.default.cloneElement(props.pathComponent, {
      d: voronoiPath,
      className: className
    }));

    return /*#__PURE__*/_react.default.cloneElement(props.groupComponent, {}, [voronoiClipPath, circle]);
  }

  return /*#__PURE__*/_react.default.cloneElement(props.pathComponent, _objectSpread(_objectSpread(_objectSpread({}, sharedProps), userProps), {}, {
    d: voronoiPath
  }));
};

exports.Voronoi = Voronoi;
