// victory-canvas/lib/canvas-group.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CanvasGroup = void 0;

var _react = _interopRequireDefault(require("react"));

var _useCanvasContext = require("./hooks/use-canvas-context");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CanvasGroup = function (props) {
  var canvasRef = _react.default.useRef(null);

  var children = props.children,
      _props$width = props.width,
      width = _props$width === void 0 ? 0 : _props$width,
      _props$height = props.height,
      height = _props$height === void 0 ? 0 : _props$height,
      clipWidth = props.clipWidth,
      padding = props.padding;

  var clear = _react.default.useCallback(function (ctx) {
    return ctx.clearRect(0, 0, width, height);
  }, [width, height]); // This needs to be called in the child component to ensure it is called after the
  // shape is drawn


  var clip = _react.default.useCallback(function (ctx) {
    var paddingRight = typeof padding === "number" ? padding : (padding === null || padding === void 0 ? void 0 : padding.right) || 0;
    var paddingLeft = typeof padding === "number" ? padding : (padding === null || padding === void 0 ? void 0 : padding.left) || 0;
    var maxClipWidth = width - paddingRight - paddingLeft;
    ctx.clearRect(width - paddingRight, 0, clipWidth ? (maxClipWidth - clipWidth) * -1 : 0, height);
  }, [width, height, padding, clipWidth]);

  return /*#__PURE__*/_react.default.createElement(_useCanvasContext.CanvasContext.Provider, {
    value: {
      canvasRef: canvasRef,
      clear: clear,
      clip: clip
    }
  }, /*#__PURE__*/_react.default.createElement("foreignObject", {
    width: width,
    height: height,
    x: 0,
    y: 0
  }, /*#__PURE__*/_react.default.createElement("canvas", {
    width: width,
    height: height,
    ref: canvasRef
  })), children);
};

exports.CanvasGroup = CanvasGroup;
CanvasGroup.role = "container";
