// tss-react/compat.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useStyles = exports.tss = exports.createMakeAndWithStyles = exports.GlobalStyles = exports.keyframes = exports.createTss = exports.createWithStyles = exports.TssCacheProvider = exports.createMakeStyles = void 0;
const makeStyles_1 = require("./makeStyles");
Object.defineProperty(exports, "createMakeStyles", { enumerable: true, get: function () { return makeStyles_1.createMakeStyles; } });
Object.defineProperty(exports, "TssCacheProvider", { enumerable: true, get: function () { return makeStyles_1.TssCacheProvider; } });
const withStyles_compat_1 = require("./withStyles_compat");
Object.defineProperty(exports, "createWithStyles", { enumerable: true, get: function () { return withStyles_compat_1.createWithStyles; } });
const tss_1 = require("./tss");
Object.defineProperty(exports, "createTss", { enumerable: true, get: function () { return tss_1.createTss; } });
/** @see <https://docs.tss-react.dev/api/keyframes> */
var react_1 = require("@emotion/react");
Object.defineProperty(exports, "keyframes", { enumerable: true, get: function () { return react_1.keyframes; } });
/** @see <https://docs.tss-react.dev/api/globalstyles> */
var GlobalStyles_1 = require("./GlobalStyles");
Object.defineProperty(exports, "GlobalStyles", { enumerable: true, get: function () { return GlobalStyles_1.GlobalStyles; } });
/** @see <https://docs.tss-react.dev/api/makestyles> */
function createMakeAndWithStyles(params) {
    return Object.assign(Object.assign({}, (0, makeStyles_1.createMakeStyles)(params)), (0, withStyles_compat_1.createWithStyles)(params));
}
exports.createMakeAndWithStyles = createMakeAndWithStyles;
exports.tss = (0, tss_1.createTss)({
    "useContext": () => ({})
}).tss;
exports.useStyles = exports.tss.create({});
