// @jbrowse/plugin-circular-view/esm/index.js
import { isAbstractMenuManager } from '@jbrowse/core/util';
import Plugin from '@jbrowse/core/Plugin';
// locals
import CircularViewF from './CircularView';
import LaunchCircularViewF from './LaunchCircularView';
// icons
import DataUsageIcon from '@mui/icons-material/DataUsage';
export default class CircularViewPlugin extends Plugin {
    constructor() {
        super(...arguments);
        this.name = 'CircularViewPlugin';
    }
    install(pluginManager) {
        CircularViewF(pluginManager);
        LaunchCircularViewF(pluginManager);
    }
    configure(pluginManager) {
        if (isAbstractMenuManager(pluginManager.rootModel)) {
            pluginManager.rootModel.appendToSubMenu(['Add'], {
                label: 'Circular view',
                icon: DataUsageIcon,
                onClick: (session) => {
                    session.addView('CircularView', {});
                },
            });
        }
    }
}
export { baseChordDisplayConfig, BaseChordDisplayModel, BaseChordDisplayComponent, } from './BaseChordDisplay';
