// @jbrowse/plugin-authentication/esm/ExternalTokenModel/model.js
import { ConfigurationReference, getConf } from '@jbrowse/core/configuration';
import { InternetAccount } from '@jbrowse/core/pluggableElementTypes/models';
import { types, getRoot } from 'mobx-state-tree';
import { ExternalTokenEntryForm } from './ExternalTokenEntryForm';
const stateModelFactory = (configSchema) => {
    return InternetAccount.named('ExternalTokenInternetAccount')
        .props({
        type: types.literal('ExternalTokenInternetAccount'),
        configuration: ConfigurationReference(configSchema),
    })
        .views(self => ({
        get validateWithHEAD() {
            return getConf(self, 'validateWithHEAD');
        },
    }))
        .actions(self => ({
        getTokenFromUser(resolve, reject) {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            const { session } = getRoot(self);
            session.queueDialog((doneCallback) => [
                ExternalTokenEntryForm,
                {
                    internetAccountId: self.internetAccountId,
                    handleClose: (token) => {
                        if (token) {
                            resolve(token);
                        }
                        else {
                            reject(new Error('user cancelled entry'));
                        }
                        doneCallback();
                    },
                },
            ]);
        },
        async validateToken(token, location) {
            if (!self.validateWithHEAD) {
                return token;
            }
            const newInit = self.addAuthHeaderToInit({ method: 'HEAD' }, token);
            const response = await fetch(location.uri, newInit);
            if (!response.ok) {
                let errorMessage;
                try {
                    errorMessage = await response.text();
                }
                catch (error) {
                    errorMessage = '';
                }
                throw new Error(`Token could not be validated — ${response.status} (${response.statusText})${errorMessage ? ` (${errorMessage})` : ''}`);
            }
            return token;
        },
    }));
};
export default stateModelFactory;
