// @jbrowse/plugin-data-management/esm/HierarchicalTrackSelectorWidget/components/tree/FavoriteTracks.js
import React from 'react';
import { Badge, Tooltip } from '@mui/material';
import { observer } from 'mobx-react';
import { makeStyles } from 'tss-react/mui';
// icons
import GradeIcon from '@mui/icons-material/Grade';
import DropdownTrackSelector from './DropdownTrackSelector';
const useStyles = makeStyles()({
    smallBadge: {
        height: 14,
    },
    margin: {
        marginRight: 10,
    },
});
const FavoriteTracks = observer(function ({ model, }) {
    const { classes } = useStyles();
    const { view, favoriteTracks } = model;
    return view ? (React.createElement(DropdownTrackSelector, { onClick: () => model.setFavoritesCounter(0), tracks: favoriteTracks, model: model, extraMenuItems: favoriteTracks.length
            ? [
                { type: 'divider' },
                {
                    label: 'Clear favorites',
                    onClick: () => model.clearFavorites(),
                },
            ]
            : [
                {
                    label: 'No favorite tracks yet',
                    onClick: () => { },
                },
            ] },
        React.createElement(Tooltip, { title: "Favorite tracks" },
            React.createElement(Badge, { classes: { badge: classes.smallBadge }, color: "secondary", anchorOrigin: {
                    vertical: 'bottom',
                    horizontal: 'right',
                }, className: classes.margin, badgeContent: model.favoritesCounter },
                React.createElement(GradeIcon, null))))) : null;
});
export default FavoriteTracks;
