// @jbrowse/core/rpc/coreRpcMethods.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CoreGetFeatureDensityStats = exports.CoreFreeResources = exports.CoreRender = exports.CoreGetFeatures = exports.CoreGetFileInfo = exports.CoreGetMetadata = exports.CoreGetFeatureDetails = exports.CoreGetRefNames = void 0;
var CoreGetRefNames_1 = require("./methods/CoreGetRefNames");
Object.defineProperty(exports, "CoreGetRefNames", { enumerable: true, get: function () { return __importDefault(CoreGetRefNames_1).default; } });
var CoreGetFeatureDetails_1 = require("./methods/CoreGetFeatureDetails");
Object.defineProperty(exports, "CoreGetFeatureDetails", { enumerable: true, get: function () { return __importDefault(CoreGetFeatureDetails_1).default; } });
var CoreGetMetadata_1 = require("./methods/CoreGetMetadata");
Object.defineProperty(exports, "CoreGetMetadata", { enumerable: true, get: function () { return __importDefault(CoreGetMetadata_1).default; } });
var CoreGetFileInfo_1 = require("./methods/CoreGetFileInfo");
Object.defineProperty(exports, "CoreGetFileInfo", { enumerable: true, get: function () { return __importDefault(CoreGetFileInfo_1).default; } });
var CoreGetFeatures_1 = require("./methods/CoreGetFeatures");
Object.defineProperty(exports, "CoreGetFeatures", { enumerable: true, get: function () { return __importDefault(CoreGetFeatures_1).default; } });
var CoreRender_1 = require("./methods/CoreRender");
Object.defineProperty(exports, "CoreRender", { enumerable: true, get: function () { return __importDefault(CoreRender_1).default; } });
var CoreFreeResources_1 = require("./methods/CoreFreeResources");
Object.defineProperty(exports, "CoreFreeResources", { enumerable: true, get: function () { return __importDefault(CoreFreeResources_1).default; } });
var CoreGetFeatureDensityStats_1 = require("./methods/CoreGetFeatureDensityStats");
Object.defineProperty(exports, "CoreGetFeatureDensityStats", { enumerable: true, get: function () { return __importDefault(CoreGetFeatureDensityStats_1).default; } });
