// material-ui-popup-state/HoverPopover.js
"use strict";

var _interopRequireDefault2 = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _extends2 = _interopRequireDefault2(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault2(require("@babel/runtime/helpers/defineProperty"));
var React = _interopRequireWildcard(require("react"));
var _material = require("@mui/material");
var _interopRequireDefault = require("./util/interopRequireDefault.js");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
// work around https://github.com/vercel/next.js/issues/57285
var Popover = (0, _interopRequireDefault.interopRequireDefault)(_material.Popover);
var HoverPopover = /*#__PURE__*/React.forwardRef(function HoverPopover(props, ref) {
  var _props$slotProps2, _props$PaperProps;
  var paperSlotProps = React.useMemo(function () {
    var _props$slotProps;
    var wrapped = (_props$slotProps = props.slotProps) === null || _props$slotProps === void 0 ? void 0 : _props$slotProps.paper;
    if (wrapped instanceof Function) {
      // eslint-disable-next-line @typescript-eslint/ban-types
      return function (ownerProps) {
        var base = wrapped(ownerProps);
        return _objectSpread(_objectSpread({}, base), {}, {
          style: _objectSpread({
            pointerEvents: 'auto'
          }, base === null || base === void 0 ? void 0 : base.style)
        });
      };
    }
    return _objectSpread(_objectSpread({}, wrapped), {}, {
      style: _objectSpread({
        pointerEvents: 'auto'
      }, wrapped === null || wrapped === void 0 ? void 0 : wrapped.style)
    });
  }, [(_props$slotProps2 = props.slotProps) === null || _props$slotProps2 === void 0 ? void 0 : _props$slotProps2.paper]);
  return /*#__PURE__*/React.createElement(Popover, (0, _extends2["default"])({}, props, {
    ref: ref,
    style: _objectSpread({
      pointerEvents: 'none'
    }, props.style),
    PaperProps: _objectSpread(_objectSpread({}, props.PaperProps), {}, {
      style: _objectSpread({
        pointerEvents: 'auto'
      }, (_props$PaperProps = props.PaperProps) === null || _props$PaperProps === void 0 ? void 0 : _props$PaperProps.style)
    }),
    slotProps: _objectSpread(_objectSpread({}, props.slotProps), {}, {
      paper: paperSlotProps
    })
  }));
});
var _default = exports["default"] = HoverPopover;
module.exports = exports.default;
